### PR TITLE
Build message enhancements

### LINKED ISSUE #43: Add separate intervals for checking progress and posting messages
The current implementation of the `!deploy_app` command checks every `n` seconds whether the build is finished or not, and posts a message every time it does so.

crates.io builds are really slow (taking around 20 minutes each), so posting a message every iteration spams the channel too much. If we increase the interval for posting messages though we risk delaying the finish of the build, as the bot would wait the same interval even for checking whether the build finished or not.

We should implement two separate intervals for this.

### PR DESCRIPTION
**Pull request by nellshamrell**
Fixes #43 

This adds in an environmental variable that allows the user to control how often build pending messages are displayed in the Discord channel SEPARATELY from how often the bot checks whether the build is completed.

I'm doing some interesting things with threads in this PR and would welcome any feedback or advice on how to make it better.

### CODE REVIEW

File: src/commands/heroku.rs [hunk 1]
```diff
@@ -414,23 +415,76 @@ pub fn deploy_app(ctx: &mut Context, msg: &Message, mut args: Args) -> CommandRe
 
     msg.reply(&ctx, build_response(&app_name, &build))?;
 
-    let mut build_pending = true;
+    let build_pending = Arc::new(Mutex::new(true));
```
Comment (pietroalbini) [inline src/commands/heroku.rs:418]: You can replace this with a [`std::sync::atomic::AtomicBool`](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicBool.html), which is better than a boolean in a mutex.

File: src/commands/heroku.rs [hunk 2]
```diff
@@ -414,23 +415,76 @@ pub fn deploy_app(ctx: &mut Context, msg: &Message, mut args: Args) -> CommandRe
...
+                }
+            }
+        },
+    ));
+
+    while *build_pending.lock().unwrap() {
         let build = heroku_client(ctx).request(&builds::BuildDetails {
             app_id: app_name.clone(),
             build_id: build.clone().id,
         })?;
```
Comment (pietroalbini) [inline src/commands/heroku.rs:480]: Did you mean to make this API call in the `build_check` scheduled job? This way you're calling Heroku two times a second.

File: src/commands/heroku.rs [hunk 3]
```diff
@@ -414,23 +415,76 @@ pub fn deploy_app(ctx: &mut Context, msg: &Message, mut args: Args) -> CommandRe
...
+    sched.add(Job::new(
+        job_interval(build_message_display_interval)
+            .parse()
+            .unwrap(),
+        move || {
+
+            // Doing manual error handling
+            // Because the try (?) operator cannot be 
+            // used in a closure
+            // (As of July 2020)
```
Comment (pietroalbini) [inline src/commands/heroku.rs:458]: Nit: the newlines in this comment seem a bit off?

File: src/commands/heroku.rs (hunk 1 above)
Comment (nellshamrell) [reply src/commands/heroku.rs]: Done!

**Comment by nellshamrell** [comment]
@pietroalbini - thanks for the review! Changes are complete.

**Comment by nellshamrell** [comment]
except...I may have introduced a bug. Will fix on Mon.

**Comment by nellshamrell** [comment]
@pietroalbini - I took the massively simpler route that you suggested - it's much better now!

File: src/commands/heroku.rs [hunk 4]
```diff
@@ -414,26 +416,114 @@ pub fn deploy_app(ctx: &mut Context, msg: &Message, mut args: Args) -> CommandRe
...
+               }
+           },
+       ));
+
+       while build_pending.load(Ordering::Relaxed) {
+           sched.tick();
+           std::thread::sleep(Duration::from_millis(500));
+       }
+
+    */
```
Comment (pietroalbini) [inline src/commands/heroku.rs:525]: Can we remove the commented code?

File: src/commands/heroku.rs [hunk 5]
```diff
@@ -631,3 +721,8 @@ fn source_url(ctx: &Context, git_sha: &str) -> String {
         git_sha,
     )
 }
+
+fn job_interval(interval: String) -> String {
```
Comment (pietroalbini) [inline src/commands/heroku.rs:725]: This function is not needed anymore.

File: src/main.rs [hunk 6]
```diff
@@ -10,6 +10,7 @@ fn main() {
         dotenv::var("DISCORD_TOKEN").expect("DISCORD_TOKEN must be set"),
         dotenv::var("HEROKU_API_KEY").expect("HEROKU_API_KEY must be set"),
         dotenv::var("BUILD_CHECK_INTERVAL").expect("BUILD_CHECK_INTERVAL must be set"),
+        dotenv::var("BUILD_MESSAGE_DISPLAY_INTERVAL").expect("BUILD_CHECK_INTERVAL must be set"),
```
Comment (pietroalbini) [inline src/main.rs:13]: The `.env.sample` needs to be updated with this variable.

Also, let's remember to also add it on ECS :)

File: Cargo.toml [hunk 7]
```diff
@@ -10,7 +10,8 @@ license = "MIT OR Apache-2.0"
 [dependencies]
 dotenv = "0.15"
 heroku_rs = { version = "0.4.1" }
+job_scheduler = "*"
```
Comment (pietroalbini) [inline Cargo.toml:13]: We can remove this dependency now.

File: src/commands/heroku.rs (hunk 4 above)
Comment (nellshamrell) [reply src/commands/heroku.rs:525]: Gah! That's what I get for declaring the PR ready at the end of the day when I was tired. Fixing!

**Comment by nellshamrell** [comment]
Deployment checklist:

- [ ] Add BUILD_MESSAGE_DISPLAY_INTERVAL to ECS

**Comment by nellshamrell** [comment]
@pietroalbini done!

**Pull request merged**